storage2/src/state/transaction.rs

```diff
@@ -0,0 +1,45 @@
+use std::collections::HashMap;
+
+use jmt::storage::{TreeReader, TreeWriter};
+
+use super::{State, StateWrite};
+
+/// Represents a transactional set of changes to a `State` fork,
+/// implemented as a RYW cache over a `State`.
+pub(crate) struct Transaction<'a, R: TreeReader + TreeWriter> {
+    // TODO: should higher-level types be used here instead of jmt types?
+    cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
+    unwritten_changes: Vec<(jmt::KeyHash, jmt::OwnedValue)>,
+    state: &'a State<'a, R>,
+}
+
+impl<'a, R: TreeReader + TreeWriter> Transaction<'a, R> {
+    pub fn new() -> Self {
+        Self {
+            cache: todo!(),
+            unwritten_changes: todo!(),
+            state: todo!(),
+        }
+    }
+
+    pub fn begin_transaction(&mut self) {
+        // Wipe the unwritten changes
+        todo!()
+    }
+
+    pub fn end_transaction(&mut self) {
+        // Write unwritten_changes to the state
+        todo!()
+    }
```

Comment:
Why are these two methods that take `Transaction` as `&mut self`? It seems to me that it would make more sense to model a transaction as a separate object whose life ends when it is committed, rather than something reusable.

Reply:
It seems to me that the API ought to look more like what `sqlx` does to model transactions. See [how a transaction is created with `Pool::begin`](https://docs.rs/sqlx/latest/sqlx/struct.Pool.html#method.begin) and [how a transaction is ended with `Transaction::commit`](https://docs.rs/sqlx/latest/sqlx/struct.Transaction.html#method.commit).